src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,939 @@
+
+    # retention (m^3/yr) * replacement cost ($/m^3) = retention value ($/yr)
+    value_array[valid_mask] = (
+        retention_volume_array[valid_mask] * replacement_cost)
+    return value_array
+
+
+def adjust_op(ratio_array, avg_ratio_array, near_impervious_lulc_array,
+              near_road_array):
+    """Apply the retention ratio adjustment algorithm to an array of ratios.
+
+    This is meant to be used with raster_calculator. Assumes that the nodata
+    value for all four input arrays is the global FLOAT_NODATA.
+
+    Args:
+        ratio_array (numpy.ndarray): 2D array of stormwater retention ratios
+        avg_ratio_array (numpy.ndarray): 2D array of averaged ratios
+        near_impervious_lulc_array (numpy.ndarray): 2D boolean array where 1
+            means this pixel is near a directly-connected LULC area
+        near_road_array (numpy.ndarray): 2D boolean array where 1
+            means this pixel is near a road centerline
+
+    Returns:
+        2D numpy array of adjusted retention ratios. Has the same shape as
+        ``retention_ratio_array``.
+    """
+    adjusted_ratio_array = numpy.full(ratio_array.shape, FLOAT_NODATA,
+                                      dtype=numpy.float32)
+    adjustment_factor_array = numpy.full(ratio_array.shape, FLOAT_NODATA,
+                                         dtype=numpy.float32)
+    valid_mask = (
+        ~numpy.isclose(ratio_array, FLOAT_NODATA) &
+        ~numpy.isclose(avg_ratio_array, FLOAT_NODATA) &
+        (near_impervious_lulc_array != UINT8_NODATA) &
+        (near_road_array != UINT8_NODATA))
+
+    # adjustment factor:
+    # - 0 if any of the nearby pixels are impervious/connected;
+    # - average of nearby pixels, otherwise
+    is_not_impervious = ~(
+        near_impervious_lulc_array[valid_mask] |
+        near_road_array[valid_mask]).astype(bool)
+    adjustment_factor_array[valid_mask] = (avg_ratio_array[valid_mask] *
+                                           is_not_impervious)
+
+    adjustment_factor_array[valid_mask] = (
+        avg_ratio_array[valid_mask] * ~(
+            near_impervious_lulc_array[valid_mask] |
+            near_road_array[valid_mask]
+        ).astype(bool))
+
+    # equation 2-4: Radj_ij = R_ij + (1 - R_ij) * C_ij
+    adjusted_ratio_array[valid_mask] = (ratio_array[valid_mask] +
+                                        (1 - ratio_array[valid_mask]) * adjustment_factor_array[valid_mask])
+    return adjusted_ratio_array
+
+
+def aggregate_results(base_aggregate_areas_path, target_path, wkt, aggregations):
+    """Aggregate outputs into regions of interest.
+
+    Args:
+        base_aggregate_areas_path (str): path to vector of polygon(s) to
+            aggregate over. This is the original input.
+        target_path (str): path to write out the results. This will be a copy
+            of the base vector with added fields, reprojected to the target WKT
+            and saved in geopackage format.
```

Comment:
I'm sorry I hadn't mentioned this before, but could I suggest something like `target_vector_path` or `target_gpkg_path` as an alternative parameter name?  That might help us more readily grok the meaning when we come back to this model in the future.

Reply:
Sure!